### PR TITLE
[rllib] Pull out multi-gpu optimizer as a generic class

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

First cut at pulling out the multi-gpu optimizer, by adding it for DQN. New classes:
 - LocalMultiGPUOptimizer
 - SampleBatch

Pong results with 42x42 and multi_gpu on and off. Perf with multi_gpu is a bit lower, but that's to be expected since we have to do a separate pass for prioritization.

```
 - DQN_PongDeterministic-v4_0_multi_gpu=True:   TERMINATED [pid=63073], 9824 s, 1270000 ts, 20 rew
 - DQN_PongDeterministic-v4_2_multi_gpu=True:   TERMINATED [pid=63462], 27332 s, 3741000 ts, 20.1 rew
 - DQN_PongDeterministic-v4_4_multi_gpu=True:   TERMINATED [pid=63475], 15352 s, 2016000 ts, 20 rew
 - DQN_PongDeterministic-v4_6_multi_gpu=True:   TERMINATED [pid=63492], 13921 s, 1815000 ts, 20 rew
 - DQN_PongDeterministic-v4_1_multi_gpu=False:  TERMINATED [pid=63211], 7237 s, 1021000 ts, 20 rew
 - DQN_PongDeterministic-v4_3_multi_gpu=False:  TERMINATED [pid=63464], 13237 s, 1909000 ts, 20 rew
 - DQN_PongDeterministic-v4_5_multi_gpu=False:  TERMINATED [pid=63485], 6648 s, 940000 ts, 20.1 rew
 - DQN_PongDeterministic-v4_7_multi_gpu=False:  TERMINATED [pid=63511], 8436 s, 1204000 ts, 20 rew
```

## Related issue number

https://github.com/ray-project/ray/issues/1291

### CODE REVIEW

**Comment by ericl** [comment]
Hm, the pong results look on par with before, but seem to be much slower. Running against the previous code to compare.

File: python/ray/rllib/optimizers/sample_batch.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from functools import reduce
+
+import numpy as np
+
+
+class SampleBatch(object):
```
Comment (ericl) [inline python/ray/rllib/optimizers/sample_batch.py]: TODO: needs to be consolidated with partial rollout somehow

File: python/ray/rllib/optimizers/sample_batch.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    """Wrapper around a dictionary with string keys and array-like values.
+
+    For example, {"obs": [1, 2, 3], "reward": [0, -1, 1]} is a batch of three
+    samples, each with an "obs" and "reward" attribute.
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/sample_batch.py]: Should Shuffle also be a method here?

File: python/ray/rllib/dqn/base_evaluator.py [hunk 3]
```diff
@@ -55,20 +55,23 @@ def update_target(self):
         self.dqn_graph.update_target(self.sess)
 
     def sample(self):
-        output = []
+        obs, actions, rewards, new_obs, dones = [], [], [], [], []
```
Comment (richardliaw) [inline python/ray/rllib/dqn/base_evaluator.py]: Does it make sense to replace this with a `SyncSampler`? A little modification would need to be made to support exploration, but this class will end up cleaner. Can be done in another PR I guess

File: python/ray/rllib/optimizers/sample_batch.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/rllib/optimizers/sample_batch.py]: So this is a nice abstraction - perhaps it makes sense to extend PartialRollout from this which provides the ability to support truncated rollouts. If so, we can do it in a separate PR.

File: python/ray/rllib/dqn/replay_evaluator.py [hunk 4]
```diff
@@ -63,38 +64,44 @@ def sample(self, no_replay=False):
...
+            for row in s.rows():
+                self.replay_buffer.add(
+                    row["obs"], row["actions"], row["rewards"], row["new_obs"],
+                    row["dones"])
 
         if no_replay:
             return samples
 
         # Then return a batch sampled from the buffer
         if self.config["prioritized_replay"]:
```
Comment (richardliaw) [inline python/ray/rllib/dqn/replay_evaluator.py]: I hope later on all the ReplayBuffer casework can get pushed into the ReplayBuffer

File: python/ray/rllib/dqn/replay_evaluator.py [hunk 5]
```diff
@@ -63,38 +64,44 @@ def sample(self, no_replay=False):
...
-                beta=self.beta_schedule.value(self.global_timestep))
             (obses_t, actions, rewards, obses_tp1,
-                dones, _, batch_idxes) = experience
+                dones, weights, batch_indexes) = self.replay_buffer.sample(
+                    self.config["train_batch_size"],
+                    beta=self.beta_schedule.value(self.global_timestep))
             self._update_priorities_if_needed()
-            self.samples_to_prioritize = (
-                obses_t, actions, rewards, obses_tp1, dones, batch_idxes)
+            batch = SampleBatch({
```
Comment (richardliaw) [inline python/ray/rllib/dqn/replay_evaluator.py]: you could just have the `replay_buffer` return a `SampleBatch` right?

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 6]
```diff
@@ -2,8 +2,103 @@
...
+        # per-GPU graph copies created below must share vars with the policy
+        tf.get_variable_scope().reuse_variables()
+
+        self.par_opt = LocalSyncParallelOptimizer(
+            tf.train.AdamOptimizer(self.config.get("sgd_stepsize", 5e-5)),
+            self.devices,
+            [ph for _, ph in self.loss_inputs],
+            self.per_device_batch_size,
+            lambda *ph: self.local_evaluator.build_tf_loss(ph),
+            self.config.get("logdir", "/tmp/ray"))
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: why this directory? didn't logging directory change per #1297

**Review by richardliaw — COMMENTED** [review]
Some comments; this introduces some nice abstractions.

File: python/ray/rllib/dqn/base_evaluator.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/dqn/base_evaluator.py]: Sure, I think it should work.

File: python/ray/rllib/dqn/replay_evaluator.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/dqn/replay_evaluator.py]: It's difficult. The prioritization requires extra computation that has to be part of the loss.

File: python/ray/rllib/dqn/replay_evaluator.py (hunk 5 above)
Comment (ericl) [reply python/ray/rllib/dqn/replay_evaluator.py]: Yes, but I'm less concerned about that for now -- I'd rather keep replay_buffer unchanged as long as possible since it's still basically similar to the baselines code.

File: python/ray/rllib/optimizers/multi_gpu.py (hunk 6 above)
Comment (ericl) [reply python/ray/rllib/optimizers/multi_gpu.py]: Changed it to os.getcwd()

File: python/ray/rllib/optimizers/sample_batch.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/optimizers/sample_batch.py]: Hmm possibly. Right now multi gpu optimizer does a more efficient shuffling where it only permutes indices, which seems just as good.

File: python/ray/rllib/optimizers/sample_batch.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        assert len(set(lengths)) == 1, "data columns must be same length"
+
+    @staticmethod
+    def concat_samples(samples):
+        return reduce(lambda a, b: a.concat(b), samples)
+
+    def concat(self, other):
+        """Returns a new SampleBatch with each data column concatenated.
+
+        >>> b1 = SampleBatch({"a": [1, 2]})
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/optimizers/sample_batch.py]: Done!

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
Jenkins is failing, but otherwise tests are fine
